ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/SnapProtocolManager.java

```diff
@@ -0,0 +1,155 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.eth.manager;
+
+import org.hyperledger.besu.ethereum.eth.SnapProtocol;
+import org.hyperledger.besu.ethereum.eth.peervalidation.PeerValidator;
+import org.hyperledger.besu.ethereum.p2p.network.ProtocolManager;
+import org.hyperledger.besu.ethereum.p2p.rlpx.connections.PeerConnection;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.AbstractSnapMessageData;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.Capability;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.Message;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.messages.DisconnectMessage;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.messages.DisconnectMessage.DisconnectReason;
+import org.hyperledger.besu.ethereum.rlp.RLPException;
+
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class SnapProtocolManager implements ProtocolManager {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final List<PeerValidator> peerValidators;
+  private final List<Capability> supportedCapabilities;
+  private final EthPeers ethPeers;
+  private final EthMessages snapMessages;
+
+  public SnapProtocolManager(
+      final List<PeerValidator> peerValidators,
+      final EthPeers ethPeers,
+      final EthMessages snapMessages) {
+    this.peerValidators = peerValidators;
+    this.ethPeers = ethPeers;
+    this.snapMessages = snapMessages;
+    this.supportedCapabilities = calculateCapabilities();
+
+    new SnapServer(snapMessages);
+  }
+
+  private List<Capability> calculateCapabilities() {
+    final ImmutableList.Builder<Capability> capabilities = ImmutableList.builder();
+    capabilities.add(SnapProtocol.SNAP1);
+
+    return capabilities.build();
+  }
+
+  @Override
+  public String getSupportedProtocol() {
+    return SnapProtocol.NAME;
+  }
+
+  @Override
+  public List<Capability> getSupportedCapabilities() {
+    return supportedCapabilities;
+  }
+
+  @Override
+  public void stop() {}
+
+  @Override
+  public void awaitStop() throws InterruptedException {}
+
+  /**
+   * This function is called by the P2P framework when an "SNAP message has been received. This
+   * function is responsible for:
+   *
+   * @param cap The capability under which the message was transmitted.
+   * @param message The message to be decoded.
+   */
+  @Override
+  public void processMessage(final Capability cap, final Message message) {
+    final MessageData messageData =
+        new AbstractSnapMessageData(message.getData().getData()) {
+          @Override
+          public int getCode() {
+            return message.getData().getCode();
+          }
+        };
```

Comment:
could we do a static delegating `create` method in AbstractSnapMessageData instead?  We won't get any of the concrete method overrides if we do an anonymous class that we may use, like for example in EthMessage.

Reply:
good idea